src/Microsoft.VisualStudio.Composition/LazyMetadataWrapper.cs

```diff
@@ -195,6 +197,32 @@ public bool Remove(KeyValuePair<string, object> item)
             return metadata;
         }
 
+        internal static bool TryGetMetadataValueTypeRef(IReadOnlyDictionary<string, object> metadata, string key, Resolver resolver, out object typeRef)
+        {
+            typeRef = null;
+
+            if (metadata is LazyMetadataWrapper lazyMetadata)
+            {
+                metadata = lazyMetadata.underlyingMetadata;
+            }
+
+            if (!metadata.TryGetValue(key, out var value))
+            {
+                return false;
+            }
+
+            if (value is ISubstitutedValue substitutedValue)
+            {
+                typeRef = substitutedValue.SubstitutedValueTypeRef;
+            }
+            else
+            {
+                typeRef = value;
+            }
```

Comment:
This could be an instance of `Type`, but it seems we should coerce it to be `TypeRef` since the caller has optimizations that only apply if it is a `TypeRef`.